<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [39]:
import pandas as pd
import psycopg2

In [40]:
pip install lxml

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.2.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [41]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

connection = psycopg2.connect(
   dbname=DBNAME,
   user=USER,
   host=HOST,
   password=PASSWORD,
   port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [42]:
# текст запроса
query_3_1 = f'''select
              COUNT(id)
           from public.VACANCIES
'''

In [43]:
# результат запроса
df = pd.read_sql_query(query_3_1 , connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\619879664.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1 , connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [44]:
# текст запроса
query_3_2 = f'''select
              COUNT(id)
           from public.EMPLOYERS
'''

In [45]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [46]:
# текст запроса
query_3_3 = f'''select
              COUNT(id)
           from public.AREAS
'''

In [47]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [48]:
# текст запроса
query_3_4 = f'''select
              COUNT(id)
           from public.INDUSTRIES
'''

In [49]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

# Выводы по предварительному анализу данных
Таким образом, мы получили следующие результаты предварительного анализа данных:
    общее количество вакансий - 49 197;
    совокупное количество работодателей - 23 591;
    общее количество регионов - 1 362;
    совокупное количество сфер деятельности - 294.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [50]:
# текст запроса
query_4_1 = f'''SELECT 
    a.name as areas_name,
    COUNT(v.id) as "Количество вакансий"
FROM   
    public.areas a
JOIN public.VACANCIES v ON v.area_id = a.id
GROUP BY areas_name
ORDER BY 2 DESC
LIMIT 5
'''

In [51]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


areas_name  Количество вакансий
0           Москва                 5333
1  Санкт-Петербург                 2851
2            Минск                 2112
3      Новосибирск                 2006
4           Алматы                 1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [52]:
# текст запроса
query_4_2 = f'''SELECT 
    COUNT(VACANCIES.id) as "Количество вакансий"
FROM   
    public.VACANCIES
    WHERE salary_from IS NOT NULL or salary_to IS NOT NULL
'''

In [53]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


Количество вакансий
0                24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [54]:
# текст запроса
query_4_3 = f'''SELECT 
    round (AVG(salary_from)) AS "средние по нижней границе заработной платы",
    round (AVG(salary_to)) AS "средние по верхней границе заработной платы"
FROM   
    public.VACANCIES
'''

In [55]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


средние по нижней границе заработ  средние по верхней границе зарабо
0                            71065.0                           110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [56]:
# текст запроса
query_4_4 = f'''SELECT 
    VACANCIES.schedule as "Тип рабочего графика",
    VACANCIES.employment as "Тип трудоустройства",
    COUNT(VACANCIES.id) as "Количество вакансий"
FROM   
    public.VACANCIES
GROUP BY VACANCIES.schedule,VACANCIES.employment
ORDER BY 3 DESC
LIMIT 2
'''

In [57]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


Тип рабочего графика Тип трудоустройства  Количество вакансий
0          Полный день    Полная занятость                35367
1     Удаленная работа    Полная занятость                 7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [58]:
# текст запроса
query_4_5 = f'''SELECT 
    VACANCIES.experience as "Опыт работы",
    COUNT(VACANCIES.id) as "Количество вакансий"
FROM   
    public.VACANCIES
GROUP BY VACANCIES.experience
ORDER BY 2 ASC
LIMIT 4
'''

In [59]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


Опыт работы  Количество вакансий
0         Более 6 лет                 1337
1           Нет опыта                 7197
2       От 3 до 6 лет                14511
3  От 1 года до 3 лет                26152

# Выводы по детальному анализу вакансий
Наибольшее количество вакансий сосредоточено в Москве и Санкт-Петербурге. Средний размер зарплатных ожиданий варьируется от 71 до 111 тыс. руб. Для наибольшего количества вакансий требуется опыт от 1 до 3-х лет

***

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [60]:
# текст запроса
query_5_1 = f'''(SELECT 
    EMPLOYERS.name as "Название работодателя",
    COUNT(VACANCIES.id) as "Количество вакансий"
FROM   
    public.EMPLOYERS
JOIN public.VACANCIES ON VACANCIES.employer_id = EMPLOYERS.id
GROUP BY EMPLOYERS.name
ORDER BY 2 DESC
LIMIT 1)
UNION ALL
(SELECT 
    EMPLOYERS.name as "Название работодателя",
    COUNT(VACANCIES.id) as "Количество вакансий"
FROM   
    public.EMPLOYERS
JOIN public.VACANCIES ON VACANCIES.employer_id = EMPLOYERS.id
GROUP BY EMPLOYERS.name
ORDER BY 2 DESC
OFFSET 4 LIMIT 1)
'''


In [61]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


Название работодателя  Количество вакансий
0                Яндекс                 1933
1         Газпром нефть                  331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [62]:
# текст запроса
query_5_2 = f'''SELECT 
    a.name as "Название региона",
    COUNT(distinct e.id) as "Количество работодателей",
    COUNT(distinct v.id) as "Количество вакансий"
FROM   
    public.AREAS a
JOIN public.EMPLOYERS e ON e.area = a.id
LEFT JOIN public.VACANCIES v ON v.area_id = a.id
WHERE v.id is NULL
GROUP BY a.name
ORDER BY 2 DESC
LIMIT 1
'''

In [63]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


Название региона  Количество работодателей  Количество вакансий
0           Россия                       410                    0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [64]:
# текст запроса
query_5_3 = f'''SELECT 
    e.name as "Название работодателя",
    COUNT(distinct v.area_id) as "Количество регионов где публикуются вакансии",
    COUNT(distinct v.id) as "Количество вакансий"
FROM   
    public.EMPLOYERS e
JOIN public.VACANCIES v ON v.employer_id = e.id
GROUP BY e.name
ORDER BY 3 DESC
LIMIT 10
'''

In [65]:
# результат запроса
df = pd.read_sql_query (query_5_3, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\2827807884.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_5_3, connection)


Название работодателя  Количество регионов где публикуют  \
0                 Яндекс                                181   
1             Ростелеком                                152   
2               Тинькофф                                 43   
3                   СБЕР                                 24   
4          Газпром нефть                                 24   
5            ИК СИБИНТЕК                                 46   
6                    МТС                                 55   
7                DataArt                                  2   
8  Совкомбанк Технологии                                 26   
9             Первый Бит                                 43   

   Количество вакансий  
0                 1933  
1                  491  
2                  444  
3                  428  
4                  331  
5                  327  
6                  292  
7                  247  
8                  204  
9                  176

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [66]:
# текст запроса
query_5_4 = f'''SELECT
    COUNT(distinct e.id) as "Количество работодателей"
FROM   
    public.EMPLOYERS e
LEFT JOIN public.EMPLOYERS_INDUSTRIES i ON e.id = i.employer_id
WHERE i.industry_id IS NULL
'''

In [67]:
# результат запроса
df = pd.read_sql_query (query_5_4, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\796163143.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_5_4, connection)


Количество работодателей
0                      8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [68]:
# текст запроса
query_5_5 = f'''SELECT
    e.name as "Название работодателя",
    COUNT(distinct i.industry_id) as "Количество сфер деятельности"
FROM   
    public.EMPLOYERS e
LEFT JOIN public.EMPLOYERS_INDUSTRIES i ON e.id = i.employer_id
GROUP BY e.name
HAVING COUNT(distinct i.industry_id) = 4
ORDER BY 1 ASC
OFFSET 2 LIMIT 1
'''

In [69]:
# результат запроса
df = pd.read_sql_query (query_5_5, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\3433731835.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_5_5, connection)


Название работодателя  Количество сфер деятельности
0                  2ГИС                             4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [70]:
# текст запроса
query_5_6 = f'''SELECT
    COUNT(distinct e.id) as "Количество работодателей"
FROM   
    public.EMPLOYERS e   
LEFT JOIN public.EMPLOYERS_INDUSTRIES i ON e.id = i.employer_id
JOIN public.INDUSTRIES y ON y.id = i.industry_id 
WHERE y.name = 'Разработка программного обеспечения'
'''

In [71]:
# результат запроса
df = pd.read_sql_query (query_5_6, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\1053093668.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_5_6, connection)


Количество работодателей
0                      3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [72]:
# код для получения списка городов-милионников_1
from urllib.parse import quote_plus
url = quote_plus('https://ru.wikipedia.org/wiki/Города-миллионеры_России', safe=':/')
tables = pd.read_html(url)
mln_cities = tables[1]['Город'].values.tolist()

In [73]:
# код для получения списка городов-милионников_2
import urllib.parse as req

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
mln_cities = tuple(pd.read_html(url)[1]['Город'])

In [74]:
# текст запроса
query_5_7 = f'''
    select
        a.name as region,
        count(*) as vacancies_count
    from
        public.vacancies as v
        join public.employers as e on v.employer_id = e.id
        join public.areas as a on v.area_id = a.id
    group by
        v.area_id, a.name, e.id, e.name
    having
        a.name in {tuple(mln_cities)} and e.name = 'Яндекс'
    

    union all
    

    select
        'Total',
        sum(x.cnt)
    from
        (
        select
            count(*) as cnt
        from
            public.vacancies as v
            join public.employers as e on v.employer_id = e.id
            join public.areas as a on v.area_id = a.id
        group by
            v.area_id, a.name, e.id, e.name
        having
            a.name in {tuple(mln_cities)} and e.name = 'Яндекс'
        ) as x
        

    order by 2
    '''

In [75]:
# результат запроса
df = pd.read_sql_query (query_5_7, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\810614778.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_5_7, connection)


region  vacancies_count
0              Омск             21.0
1         Челябинск             22.0
2        Красноярск             23.0
3         Волгоград             24.0
4    Ростов-на-Дону             25.0
5            Казань             25.0
6             Пермь             25.0
7               Уфа             26.0
8            Самара             26.0
9         Краснодар             30.0
10          Воронеж             32.0
11      Новосибирск             35.0
12  Нижний Новгород             36.0
13     Екатеринбург             39.0
14  Санкт-Петербург             42.0
15           Москва             54.0
16            Total            485.0

***

# Выводы по анализу работодателей
Наибольшее количество вакансий в базе - у компании Яндекс, упри этом среди городов-миллионников максимум вакансий приходятся на Москву, Санкт-Петербург и Екатеринбург

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [76]:
# текст запроса
query_6_1 = '''select count(*)
                from public.vacancies
                where name ilike '%data%' or name ilike '%данн%'
'''

In [77]:
# результат запроса
df = pd.read_sql_query (query_6_1, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\2916403832.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [78]:
# текст запроса
query_6_2 = '''select count(*)
    from
        public.vacancies v
    where
        (v.name ILIKE '%data scientist%' OR
        v.name ILIKE '%data science%'OR
        v.name ILIKE '%исследователь данных%' OR
        (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
        v.name ILIKE '%machine learning%' OR v.name ILIKE '%машинн%обучен%')
        and
        (v.name ILIKE '%junior%' or
        v.experience = 'Нет опыта' or
        v.employment ILIKE '%стажировка%')
'''

In [79]:
# результат запроса
df = pd.read_sql_query (query_6_2, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\744361491.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [80]:
# текст запроса
query_6_3 = f'''
select count(*)
    from
        public.vacancies
    where
        (
        lower(name) like '%data scientist%'  or
        lower(name) like '%data science%' or
        lower(name) like '%исследователь данных%'  or
        (name like '%ML%' and name not ilike '%HTML%')  or
        lower(name) like '%machine learning%'  or
        lower(name) like '%машинн%обучен%'
        )
        and
        (
        key_skills ilike '%SQL%' or
        key_skills like '%postgres%'
        )
'''

In [81]:
# результат запроса
df = pd.read_sql_query (query_6_3, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\2436245350.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [82]:
# текст запроса
query_6_4 = f'''
select count(*)
    from
        public.vacancies
    where
        (
        lower(name) like '%data scientist%'  or
        lower(name) like '%data science%' or
        lower(name) like '%исследователь данных%'  or
        (name like '%ML%' and name not ilike '%HTML%')  or
        lower(name) like '%machine learning%'  or
        lower(name) like '%машинн%обучен%'
        )
        and
        key_skills ilike '%Python%'
'''

In [83]:
# результат запроса
df = pd.read_sql_query (query_6_4, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\4166993872.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [84]:
# текст запроса
query_6_5 = f'''
SELECT 
    round(avg(length(key_skills)-length(replace(key_skills, CHR(9), ''))+1), 2) as result
FROM public.VACANCIES v
where 
(
  name ilike '%data scientist%'or
  name ilike '%data science%'or
  name ilike '%исследователь данных%'or
  (name like '%ML%' and name not ilike '%HTML%') or
  name ilike '%machine learning%' or name ilike '%машинн%обучен%'
)
'''

In [85]:
# результат запроса
df = pd.read_sql_query (query_6_5, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\1144307522.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_5, connection)


result
0    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [86]:
# текст запроса
query_6_6 = f'''
SELECT 
    experience as "опыт работы",
    ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to)), 0) as "средний размер зарплаты"
FROM public.VACANCIES v
where 
(
  name ilike '%data scientist%'or
  name ilike '%data science%'or
  name ilike '%исследователь данных%'or
  (name like '%ML%' and name not ilike '%HTML%') or
  name ilike '%machine learning%' or name ilike '%машинн%обучен%'
)
AND experience = 'От 3 до 6 лет'
AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY experience
'''

In [87]:
# результат запроса
df = pd.read_sql_query (query_6_6, connection)
df

C:\Users\Евгений\AppData\Local\Temp\ipykernel_5708\3310421505.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_6, connection)


опыт работы  средний размер зарплаты
0  От 3 до 6 лет                 243115.0

***

# Общий вывод по проекту

Анализ собранных сведений по рынку труда позволил определить, как с помощью этой базы данных можно установить количественные и качественные показатели разных видов информации, связанных с вакансиями, компаниями-работодателями, регионами и отраслями. На основе данных о вакансиях и работодателях демонстрируется возможность обнаружения присущих им закономерностей и трендов, а также показывается, как можно найти сведения по узким и специализированным выборкам.